### PR TITLE
Dbeaver/pro#3360 Fix datasets opening

### PR DESCRIPTION
**Pull request by E1izabeth**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java [hunk 1]
```diff
@@ -2292,6 +2277,35 @@ protected void doSetInput(IEditorInput editorInput) throws CoreException {
         editorImage = new Image(Display.getCurrent(), baseEditorImage, SWT.IMAGE_COPY);
     }
 
+    @NotNull
+    private Runnable getEditorInputInitializer(@NotNull IEditorInput editorInput) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java]: Let's not make it return a `Runnable`. Instead, make it do what its `run` method does, and at the call site wrap it into lambda.

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java [hunk 2]
```diff
@@ -2258,25 +2261,7 @@ protected void doSetInput(IEditorInput editorInput) throws CoreException {
...
-            }
-
-            {
-                DBPDataSourceContainer dataSource = EditorUtils.getInputDataSource(finalEditorInput);
-                SQLEditorFeatures.SQL_EDITOR_OPEN.use(Map.of(
-                    "driver", dataSource == null ? "" : dataSource.getDriver().getPreconfiguredId()
-                ));
-            }
-        };
+        Runnable inputinitializer = getEditorInputInitializer(editorInput);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java]: A continuation of the previous comment
```suggestion
        Runnable inputinitializer = () -> getEditorInputInitializer(editorInput);
```
But be careful as you might want to capture just the variable.

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java (hunk 1 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java]: I need Runnable for UIExecutionQueue.queueExec

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java]: Right, but is it the method's responsibility to make you a `Runnable`? I don't think so

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java [hunk 3]
```diff
@@ -3268,17 +3279,19 @@ public void handleDataSourceEvent(final DBPEvent event) {
...
-                : execContext.getContextDefaults();
-            if (ctxDefault != null
-                && (event.getObject() == ctxDefault.getDefaultCatalog() || event.getObject() == ctxDefault.getDefaultSchema())
-            ) {
-                contextChanged = true;
+            if (execContext != null) {
+                DBCExecutionContextDefaults<?, ?> ctxDefault = execContext.getContextDefaults();
+                if (ctxDefault != null && eventObject != null) {
+                    boolean defaultChanged = eventObject == ctxDefault.getDefaultCatalog() || eventObject == ctxDefault.getDefaultSchema();
+                    if (lastExecutionContext != executionContext || (lastExecutionContext == executionContext && defaultChanged)) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java]: Can be simplified
```suggestion
                    if (lastExecutionContext != executionContext || defaultChanged)) {
```

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java [hunk 4]
```diff
@@ -2292,6 +2287,23 @@ protected void doSetInput(IEditorInput editorInput) throws CoreException {
         editorImage = new Image(Display.getCurrent(), baseEditorImage, SWT.IMAGE_COPY);
     }
 
+    private void accomplishEditorInputInitialization(@NotNull IEditorInput editorInput) {
+        DBPDataSourceContainer oldDataSource = SQLEditor.this.getDataSourceContainer();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java:2287]: Why do you use explicit `this` everywhere?

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditor.java:2287]: It doesn't work without `this` here.
Also I like the general approach of clearly specifying that the object comes from the current class scope

**Review by ShadelessFox — APPROVED** [review]

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgreExecutionContext.java [hunk 5]
```diff
@@ -111,6 +111,10 @@ void setDefaultCatalog(@NotNull DBRProgressMonitor monitor, @NotNull PostgreData
                     disconnect();
                     setOwnerInstance(catalog);
                     connect(monitor, null, null, null, false);
+                    // Catalog has been changed. Get the new one and propagate isolated flag
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgreExecutionContext.java]: Don't understand this logic. We already have new instance (it is catalog). Why read it again why set isolated flag?

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgreExecutionContext.java]: It's the same situation as here https://github.com/dbeaver/dbeaver/pull/36404/files#diff-39b703e635479b90005852d1883628ef40d44433da8992522bd6bea810ea6ffbR125-R128
After disconnecting and connecting, the catalog is a different object, and it loses the isolatestContext property value.

**Pull request merged**